### PR TITLE
[dynamo] simplify implementation for `functools.reduce`

### PR DESCRIPTION
**Pull request by XuehaiPan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133779
* __->__ #133778

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec @xmfan

### CODE REVIEW

**Review by jansel — APPROVED** [review]

**Comment by ZainRizvi** [comment]
@pytorchbot revert -c "ghfirst" -m "Sorry, but this breaks internal tests because of using functools"

Exact error:

```
Unsupported: Graph break due to unsupported builtin _functools.reduce. This function is either a Python builtin (e.g. _warnings.warn) or a third-party C/C++ Python extension (perhaps created with pybind). If it is a Python builtin, please file an issue on GitHub so the PyTorch team can add support for it and see the next case for a workaround. If it is a third-party C/C++ Python extension, please either wrap it into a PyTorch-understood custom operator (see [https://pytorch.org/tutorials/advanced/custom_ops_landing_page.html](https://pytorch.org/tutorials/advanced/custom_ops_landing_page.html) for more details) or, if it is traceable, use torch.compiler.allow_in_graph.
```

**Comment by ZainRizvi** [comment]
@jansel can you please help @XuehaiPan test his fixes?  The corresponding diff is D61717934

**Review by anijain2305 — APPROVED** [review]

**Comment by ZainRizvi** [comment]
@pytrochbot revert -c ghfirst -m "This is still failing internally with the same error about "Graph break due to unsupported builtin _functools.reduce'"

@jansel, I would recommend helping @XuehaiPan by doing a ghimport of this diff to validate that it passes internal tests before this is relanded. 

Error:
torch/_dynamo/variables/functions.py:730: UserWarning: Graph break due to unsupported builtin _functools.reduce. This function is either a Python builtin (e.g. _warnings.warn) or a third-party C/C++ Python extension (perhaps created with pybind). If it is a Python builtin, please file an issue on GitHub so the PyTorch team can add support for it and see the next case for a workaround. If it is a third-party C/C++ Python extension, please either wrap it into a PyTorch-understood custom operator (see [https://pytorch.org/.../adva.../custom_ops_landing_page.html](https://pytorch.org/tutorials/advanced/custom_ops_landing_page.html) for more details) or, if it is traceable, use torch.compiler.allow_in_graph.
  torch._dynamo.utils.warn_once(msg)

**Comment by ZainRizvi** [comment]
@pytorchbot revert -c ghfirst -m "This is still failing internally with the same error about "Graph break due to unsupported builtin _functools.reduce'"

@jansel, I would recommend helping @XuehaiPan by doing a ghimport of this diff to validate that it passes internal tests before this is relanded.

Error:
torch/_dynamo/variables/functions.py:730: UserWarning: Graph break due to unsupported builtin _functools.reduce. This function is either a Python builtin (e.g. _warnings.warn) or a third-party C/C++ Python extension (perhaps created with pybind). If it is a Python builtin, please file an issue on GitHub so the PyTorch team can add support for it and see the next case for a workaround. If it is a third-party C/C++ Python extension, please either wrap it into a PyTorch-understood custom operator (see [https://pytorch.org/.../adva.../custom_ops_landing_page.html](https://pytorch.org/tutorials/advanced/custom_ops_landing_page.html) for more details) or, if it is traceable, use torch.compiler.allow_in_graph.
torch._dynamo.utils.warn_once(msg)

**Comment by ZainRizvi** [comment]
@pytorchbot revert -c ghfirst -m "This is still failing internally with the same error about 'Graph break due to unsupported builtin _functools.reduce'"

@jansel, I would recommend helping @XuehaiPan by doing a ghimport of this diff to validate that it passes internal tests before this is relanded.

Error:
torch/_dynamo/variables/functions.py:730: UserWarning: Graph break due to unsupported builtin _functools.reduce. This function is either a Python builtin (e.g. _warnings.warn) or a third-party C/C++ Python extension (perhaps created with pybind). If it is a Python builtin, please file an issue on GitHub so the PyTorch team can add support for it and see the next case for a workaround. If it is a third-party C/C++ Python extension, please either wrap it into a PyTorch-understood custom operator (see [https://pytorch.org/.../adva.../custom_ops_landing_page.html](https://pytorch.org/tutorials/advanced/custom_ops_landing_page.html) for more details) or, if it is traceable, use torch.compiler.allow_in_graph.
torch._dynamo.utils.warn_once(msg)

**Comment by ZainRizvi** [comment]
details on the failure are in D62009011

File: torch/_dynamo/variables/builtin.py [hunk 1]
```diff
@@ -1581,7 +1581,10 @@ def call_sum(self, tx: "InstructionTranslator", seq, start=_SENTINEL):
             if start is self._SENTINEL:
                 start = variables.ConstantVariable.create(0)
             items = seq.unpack_var_sequence(tx)
-            return BuiltinVariable(functools.reduce).call_function(
```
Comment (XuehaiPan) [inline torch/_dynamo/variables/builtin.py:1584]: We need to change some hardcoded handlers to use `SourcelessBuilder.create(tx, value)`.

**Review by jansel — CHANGES_REQUESTED** [review]
Please add a test that reproduces the error message that caused the revert.

File: torch/_dynamo/trace_rules.py [hunk 2]
```diff
@@ -3476,7 +3474,7 @@ def check_verbose(obj, is_inlined_call=False):
     # Consulte the central trace rules defined in torch._dynamo.trace_rules.
     reasons: Set[str] = set()
     rule = lookup_inner(fi.py_obj, fi.name, fi.filename, is_inlined_call, reasons)
-    if issubclass(rule, (UserFunctionVariable, PolyfilledFunctionVariable)):
+    if issubclass(rule, UserFunctionVariable):
```
Comment (XuehaiPan) [inline torch/_dynamo/trace_rules.py:3477]: I think the internal breakage is caused by this.

**Review by jansel — CHANGES_REQUESTED** [review]
Test failures?

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**